### PR TITLE
Adding exact search

### PR DESCRIPTION
**Pull request by Memeriaj**
[Task](https://app.asana.com/0/0/1200903621308177/f)

Allow searching for exact matches in addition to prefix matches.  Very similar only we won't progress down the tree after doing the initial node search (besides potentially visiting the empty string child).

### CODE REVIEW

File: test/trie.test.ts [hunk 1]
```diff
@@ -334,4 +334,128 @@ describe('Trie', () => {
...
+      expect(t.exactSearch('')).toEqual([]);
+      expect(t.exactSearch('a')).toEqual(['a']);
+      expect(t.exactSearch('aa')).toEqual(['aa']);
+
+      t.remove({ key: 'a' });
+      expect(t.exactSearch('a')).toEqual([]);
+      expect(t.exactSearch('aa')).toEqual(['aa']);
+    });
+
+    it('can replace values', () => {
```
Comment (Memeriaj) [inline test/trie.test.ts:446]: Any other tests we should add?  I added most of the ones that the prefix search was using for completeness.

File: src/node.ts [hunk 2]
```diff
@@ -243,6 +243,39 @@ class Node<T> {
...
+   * contains) for exact matched of the query.
+   *
+   * We use the passed in pqueue which has a limit and unique flag to
+   * configure the search.
+   */
+  getSortedExactResults(query: string, results: Array<T>, opts: SearchOptions) {
+    const seenKeys = new Set<string>();
+
+    if (!this.leaf) {
+      this.children['']?.getSortedExactResults(query, results, opts);
```
Comment (Memeriaj) [inline src/node.ts:257]: Doing a bit of recursion here.  I wasn't too concerned about doing so only because this should only happen a single time.

File: src/index.ts [hunk 3]
```diff
@@ -83,6 +83,34 @@ class Trie<T> {
...
+   * Returns an array of values that have keys matching the query exactly.
+   *
+   * You are encouraged to pass an options object with:
+   * a .limit to limit the number of results returned.
+   * a .unique property if you only want one result per-key.
+   *
+   * The limit is particularly important because the performance of the
+   * algorithm is determined primarily by the limit.
+   */
+  exactSearch(query: string, opts?: Partial<SearchOptions>): T[] {
```
Comment (Memeriaj) [inline src/index.ts:97]: I considered refactoring part of this out with a helper that would also match things with `prefixSearch`. But I think it would have only been confusing.

File: src/node.ts (hunk 2 above)
Comment (rockwotj) [reply src/node.ts:257]: Sounds good. Can you expand on the validateInvarients function to check that if the child's key is the empty string that all the children nodes are leaves?

File: src/index.ts [hunk 4]
```diff
@@ -83,6 +83,34 @@ class Trie<T> {
...
+
+  /**
+   * Returns an array of values that have keys matching the query exactly.
+   *
+   * You are encouraged to pass an options object with:
+   * a .limit to limit the number of results returned.
+   * a .unique property if you only want one result per-key.
+   *
+   * The limit is particularly important because the performance of the
+   * algorithm is determined primarily by the limit.
```
Comment (rockwotj) [inline src/index.ts:95]: This isn't quite as true here as it is above - it'd suggest removing it.

File: test/trie.test.ts (hunk 1 above)
Comment (rockwotj) [reply test/trie.test.ts:446]: I would add this to the randomized tests (it's in another file)

**Review by rockwotj — CHANGES_REQUESTED** [review]
In general the code looks good. I do have a question for you about **why** this is useful. Let's chat when you're free.

File: src/index.ts [hunk 5]
```diff
@@ -83,6 +83,34 @@ class Trie<T> {
 
     return results;
   }
+
+  /**
+   * Returns an array of values that have keys matching the query exactly.
+   *
+   * You are encouraged to pass an options object with:
+   * a .limit to limit the number of results returned.
+   * a .unique property if you only want one result per-key.
```
Comment (ryanpbrewster) [inline src/index.ts:92]: Is the `unique` property relevant for exact matches? How would one go about adding the same entry twice with the same key?

File: src/node.ts [hunk 6]
```diff
@@ -243,6 +243,39 @@ class Node<T> {
...
 
+  /*
+   * Look at only the results from this node (or the empty string node this one
+   * contains) for exact matched of the query.
+   *
+   * We use the passed in pqueue which has a limit and unique flag to
+   * configure the search.
+   */
+  getSortedExactResults(query: string, results: Array<T>, opts: SearchOptions) {
+    const seenKeys = new Set<string>();
```
Comment (ryanpbrewster) [inline src/node.ts:254]: can we move this declaration down below (closer to where it's used)?

File: src/node.ts [hunk 7]
```diff
@@ -243,6 +243,39 @@ class Node<T> {
...
+      return;
+    }
+
+    if (!this.sorted) {
+      this.sort();
+    }
+
+    for (let i = 0; i < this.values.length; i++) {
+      const item = this.values[i] as Item<T>;
+      if (item.key === query) {
```
Comment (ryanpbrewster) [inline src/node.ts:267]: Can we add a comment explaining why this isn't guaranteed? ((My assumption is that it has something to do with compressed edges))

**Review by ryanpbrewster — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 5 above)
Comment (rockwotj) [reply src/index.ts:92]: The trie does not enforce unique keys. You can have multiple Andrew Lee contacts for example.

Comment (ryanpbrewster) [reply src/index.ts:92]: Let me rephrase. Given that we're doing an exact match here, every key will necessarily be equivalent. Thus, isn't passing in `unique: true` equivalent to passing in `limit: 1`?

Comment (rockwotj) [reply src/index.ts:92]: I think that distinct is not useful here in **our** use, but the following is valid for the trie (duplicates can be inserted).

```
trie.add({key: 'andrew lee', distinct: 'andrew@a.com', score: 1});
trie.add({key: 'andrew lee', distinct: 'andrew@a.com', score: 2});
trie.add({key: 'andrew lee', distinct: 'drew@b.com', score: 3});
trie.add({key: 'andrew lee', distinct: 'lee@c.com', score: 4});
```

trie.{exact,prefix}Search({key: 'andrew lee', distinct: true}) => 3 results;
trie.{exact,prefix}Search({key: 'andrew lee', distinct: false}) => 4 results;

Comment (ryanpbrewster) [reply src/index.ts:92]: Ah, I misunderstood the naming conventions here.

- `distinct` is an optional field on _every item_.
- `key` is overloaded to mean both "the string that is used to look up this item" and "this item's `distinct` field, if present, else its normal key"
- `unique` is a search-time option to deduplicate by `key` in the second sense

When the documentation for this method says

>   * a .unique property if you only want one result per-key.

I was interpreting it to mean "one result per string used to look up this item". Indeed, if you do not provide a `distinct` field for your items, the `unique` option seems pretty useless for an exact search.

Comment (ryanpbrewster) [reply src/index.ts:92]: Even if it's just informally, I think we should have distinct words for `node.key` and `node.distinct || node.key`. I'd propose "key" and "identifier".

Comment (rockwotj) [reply src/index.ts:92]: > Indeed, if you do not provide a distinct field for your items, the unique option seems pretty useless for an exact search.

This is not true because the trie allows duplicate keys and values. So without distinct can give you the same value multiple times.

FWIW we don't use/need any of this so we can simplify the interface. I don't really care but I request to review the PRs if we do want to change them.

Comment (ryanpbrewster) [reply src/index.ts:92]: Talked out-of-band to resolve this. The documentation is imprecise and we can probably improve it, but the core insight is that `unique` operates on the `distinct` field, and is thus non-trivial.

**Comment by Memeriaj** [comment]
Contact suggestions haven't needed exact search or at least we haven't noticed it missing.  We can resurrect this PR if we decide we want to add it later.
